File: dotnet/CardsProject/Controllers/Tournaments/MatchController.cs
using Microsoft.AspNetCore.Mvc;
using System.ComponentModel.DataAnnotations;
using CardsProject.Services.Tournaments;

namespace CardsProject.Controllers.Tournaments;

[ApiController]
[Route("api/matches")]
public class MatchController : ControllerBase
{
    private readonly MatchService _svc;

    public MatchController(MatchService svc) => _svc = svc;

    [Microsoft.AspNetCore.Authorization.AllowAnonymous]
    [HttpGet]
    public async Task<IActionResult> List()
    {
        var items = await _svc.GetAllAsync();
        return Ok(items);
    }

    [Microsoft.AspNetCore.Authorization.AllowAnonymous]
    [HttpPost]
    public async Task<IActionResult> Create([FromBody] MatchDto dto)
    {
        if (!ModelState.IsValid) return BadRequest(ModelState);
        try
        {
            var entity = await _svc.CreateAsync(dto);
            return CreatedAtAction(nameof(Show), new { id = entity.Id }, entity);
        }
        catch (ValidationException ex) { return BadRequest(new { error = ex.Message }); }
        catch (InvalidOperationException ex) { return BadRequest(new { error = ex.Message }); }
        catch (Microsoft.EntityFrameworkCore.DbUpdateException ex) { return BadRequest(new { error = ex.InnerException?.Message ?? ex.Message }); }
    }

    [Microsoft.AspNetCore.Authorization.AllowAnonymous]
    [HttpGet("{id:int}")]
    public async Task<IActionResult> Show(int id)
    {
        var entity = await _svc.GetByIdAsync(id);
        if (entity is null) return NotFound();
        return Ok(entity);
    }

    [HttpPost("{id:int}/record")]
    public async System.Threading.Tasks.Task<IActionResult> RecordResult(int id, [FromBody] System.Collections.Generic.Dictionary<string, object> body)
    {
        try
        {
            var p1Wins = (int)body["p1_wins"];
            var p2Wins = (int)body["p2_wins"];
            await _svc.RecordResultAsync(id, p1Wins, p2Wins);
            return NoContent();
        }
        catch (ArgumentException ex) { return UnprocessableEntity(new { error = ex.Message }); }
        catch (KeyNotFoundException) { return NotFound(); }
    }

    [HttpPost("{id:int}/finalize")]
    public async System.Threading.Tasks.Task<IActionResult> FinalizeResult(int id)
    {
        try
        {
            await _svc.FinalizeResultAsync(id);
            return NoContent();
        }
        catch (ArgumentException ex) { return UnprocessableEntity(new { error = ex.Message }); }
        catch (KeyNotFoundException) { return NotFound(); }
    }

    [HttpGet("{id:int}/winner")]
    public async System.Threading.Tasks.Task<IActionResult> DetermineWinner(int id)
    {
        try
        {
            var result = await _svc.DetermineWinnerAsync(id);
            return Ok(result);
        }
        catch (ArgumentException ex) { return UnprocessableEntity(new { error = ex.Message }); }
        catch (KeyNotFoundException) { return NotFound(); }
    }

    [HttpPost("{id:int}/concede")]
    public async System.Threading.Tasks.Task<IActionResult> Concede(int id, [FromBody] System.Collections.Generic.Dictionary<string, object> body)
    {
        try
        {
            var playerId = (int)body["player_id"];
            await _svc.ConcedeAsync(id, playerId);
            return NoContent();
        }
        catch (ArgumentException ex) { return UnprocessableEntity(new { error = ex.Message }); }
        catch (KeyNotFoundException) { return NotFound(); }
    }

    [HttpPost("{id:int}/draw")]
    public async System.Threading.Tasks.Task<IActionResult> Draw(int id)
    {
        try
        {
            await _svc.DrawAsync(id);
            return NoContent();
        }
        catch (ArgumentException ex) { return UnprocessableEntity(new { error = ex.Message }); }
        catch (KeyNotFoundException) { return NotFound(); }
    }

    [Microsoft.AspNetCore.Authorization.Authorize(Roles = "Judge,HeadJudge,Admin")]
    [HttpPatch("{id:int}/transitions/pending-to-active")]
    public async Task<IActionResult> TransitionPendingToActive(int id)
    {
        try { return Ok(await _svc.TransitionPendingToActiveAsync(id)); }
        catch (InvalidOperationException ex) { return Conflict(new { error = ex.Message }); }
        catch (ArgumentException ex) { return UnprocessableEntity(new { error = ex.Message }); }
        catch (KeyNotFoundException) { return NotFound(); }
    }

    [Microsoft.AspNetCore.Authorization.Authorize(Roles = "Judge,HeadJudge,Admin")]
    [HttpPatch("{id:int}/transitions/active-to-completed")]
    public async Task<IActionResult> TransitionActiveToCompleted(int id)
    {
        try { return Ok(await _svc.TransitionActiveToCompletedAsync(id)); }
        catch (InvalidOperationException ex) { return Conflict(new { error = ex.Message }); }
        catch (ArgumentException ex) { return UnprocessableEntity(new { error = ex.Message }); }
        catch (KeyNotFoundException) { return NotFound(); }
    }

    [Microsoft.AspNetCore.Authorization.Authorize(Roles = "Judge,HeadJudge,Admin")]
    [HttpPatch("{id:int}/transitions/active-to-draw")]
    public async Task<IActionResult> TransitionActiveToDraw(int id)
    {
        try { return Ok(await _svc.TransitionActiveToDrawAsync(id)); }
        catch (InvalidOperationException ex) { return Conflict(new { error = ex.Message }); }
        catch (ArgumentException ex) { return UnprocessableEntity(new { error = ex.Message }); }
        catch (KeyNotFoundException) { return NotFound(); }
    }

    [Microsoft.AspNetCore.Authorization.Authorize(Roles = "Judge,HeadJudge,Admin")]
    [HttpPatch("{id:int}/transitions/pending-to-bye")]
    public async Task<IActionResult> TransitionPendingToBYE(int id)
    {
        try { return Ok(await _svc.TransitionPendingToBYEAsync(id)); }
        catch (InvalidOperationException ex) { return Conflict(new { error = ex.Message }); }
        catch (ArgumentException ex) { return UnprocessableEntity(new { error = ex.Message }); }
        catch (KeyNotFoundException) { return NotFound(); }
    }

    [HttpPatch("{id:int}/transitions/completed-to-active")]
    public async Task<IActionResult> TransitionCompletedToActive(int id)
    {
        try { return Ok(await _svc.TransitionCompletedToActiveAsync(id)); }
        catch (InvalidOperationException ex) { return Conflict(new { error = ex.Message }); }
        catch (ArgumentException ex) { return UnprocessableEntity(new { error = ex.Message }); }
        catch (KeyNotFoundException) { return NotFound(); }
    }

    [HttpPatch("{id:int}/transitions/draw-to-active")]
    public async Task<IActionResult> TransitionDrawToActive(int id)
    {
        try { return Ok(await _svc.TransitionDrawToActiveAsync(id)); }
        catch (InvalidOperationException ex) { return Conflict(new { error = ex.Message }); }
        catch (ArgumentException ex) { return UnprocessableEntity(new { error = ex.Message }); }
        catch (KeyNotFoundException) { return NotFound(); }
    }

    [HttpPatch("{id:int}/transitions/bye-to-active")]
    public async Task<IActionResult> TransitionBYEToActive(int id)
    {
        try { return Ok(await _svc.TransitionBYEToActiveAsync(id)); }
        catch (InvalidOperationException ex) { return Conflict(new { error = ex.Message }); }
        catch (ArgumentException ex) { return UnprocessableEntity(new { error = ex.Message }); }
        catch (KeyNotFoundException) { return NotFound(); }
    }
}

File: spring/src/main/java/cardsproject/controller/tournaments/MatchController.java
package cardsproject.controller.tournaments;

import cardsproject.domain.tournaments.Match;
import cardsproject.service.tournaments.MatchService;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.util.List;

@RestController
@RequestMapping("/api/matches")
public class MatchController {

    private final MatchService service;

    public MatchController(MatchService service) {
        this.service = service;
    }


    @GetMapping
    public List<Match> list() {
        return service.findAll();
    }

    @PostMapping
    public ResponseEntity<Match> create(@Valid @RequestBody Match entity) {
        return ResponseEntity.status(201).body(service.save(entity));
    }

    @GetMapping("/{id}")
    public ResponseEntity<Match> show(@PathVariable Long id) {
        return service.findById(id)
            .map(ResponseEntity::ok)
            .orElse(ResponseEntity.notFound().build());
    }


    @PostMapping("/{id}/record")
    public ResponseEntity<Void> recordResult(@PathVariable Long id, @RequestBody java.util.Map<String,Object> body) {
        try {
            service.recordResult(id, (Integer) body.get("p1_wins"), (Integer) body.get("p2_wins"));
            return ResponseEntity.noContent().build();
        } catch (IllegalStateException e) {
            return ResponseEntity.status(409).body(null);
        } catch (RuntimeException e) {
            return ResponseEntity.notFound().build();
        }
    }

    @PostMapping("/{id}/finalize")
    public ResponseEntity<Void> finalizeResult(@PathVariable Long id) {
        try {
            service.finalizeResult(id);
            return ResponseEntity.noContent().build();
        } catch (IllegalStateException e) {
            return ResponseEntity.status(409).body(null);
        } catch (RuntimeException e) {
            return ResponseEntity.notFound().build();
        }
    }

    @GetMapping("/{id}/winner")
    public ResponseEntity<Boolean> determineWinner(@PathVariable Long id) {
        try {
            return ResponseEntity.ok(service.determineWinner(id));
        } catch (IllegalStateException e) {
            return ResponseEntity.status(409).body(null);
        } catch (RuntimeException e) {
            return ResponseEntity.notFound().build();
        }
    }

    @PostMapping("/{id}/concede")
    public ResponseEntity<Void> concede(@PathVariable Long id, @RequestBody java.util.Map<String,Object> body) {
        try {
            service.concede(id, (Integer) body.get("player_id"));
            return ResponseEntity.noContent().build();
        } catch (IllegalStateException e) {
            return ResponseEntity.status(409).body(null);
        } catch (RuntimeException e) {
            return ResponseEntity.notFound().build();
        }
    }

    @PostMapping("/{id}/draw")
    public ResponseEntity<Void> draw(@PathVariable Long id) {
        try {
            service.draw(id);
            return ResponseEntity.noContent().build();
        } catch (IllegalStateException e) {
            return ResponseEntity.status(409).body(null);
        } catch (RuntimeException e) {
            return ResponseEntity.notFound().build();
        }
    }

    @org.springframework.security.access.prepost.PreAuthorize("hasAnyRole('ROLE_JUDGE')")
    @PatchMapping("/{id}/transitions/pending-to-active")
    public ResponseEntity<?> transitionPendingToActive(@PathVariable Long id) {
        try {
            return ResponseEntity.ok(service.transitionPendingToActive(id));
        } catch (IllegalStateException e) {
            return ResponseEntity.status(409).body(java.util.Map.of("error", e.getMessage()));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.unprocessableEntity().body(java.util.Map.of("error", e.getMessage()));
        } catch (RuntimeException e) {
            return ResponseEntity.notFound().build();
        }
    }

    @org.springframework.security.access.prepost.PreAuthorize("hasAnyRole('ROLE_JUDGE')")
    @PatchMapping("/{id}/transitions/active-to-completed")
    public ResponseEntity<?> transitionActiveToCompleted(@PathVariable Long id) {
        try {
            return ResponseEntity.ok(service.transitionActiveToCompleted(id));
        } catch (IllegalStateException e) {
            return ResponseEntity.status(409).body(java.util.Map.of("error", e.getMessage()));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.unprocessableEntity().body(java.util.Map.of("error", e.getMessage()));
        } catch (RuntimeException e) {
            return ResponseEntity.notFound().build();
        }
    }

    @org.springframework.security.access.prepost.PreAuthorize("hasAnyRole('ROLE_JUDGE')")
    @PatchMapping("/{id}/transitions/active-to-draw")
    public ResponseEntity<?> transitionActiveToDraw(@PathVariable Long id) {
        try {
            return ResponseEntity.ok(service.transitionActiveToDraw(id));
        } catch (IllegalStateException e) {
            return ResponseEntity.status(409).body(java.util.Map.of("error", e.getMessage()));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.unprocessableEntity().body(java.util.Map.of("error", e.getMessage()));
        } catch (RuntimeException e) {
            return ResponseEntity.notFound().build();
        }
    }

    @org.springframework.security.access.prepost.PreAuthorize("hasAnyRole('ROLE_JUDGE')")
    @PatchMapping("/{id}/transitions/pending-to-bye")
    public ResponseEntity<?> transitionPendingToBYE(@PathVariable Long id) {
        try {
            return ResponseEntity.ok(service.transitionPendingToBYE(id));
        } catch (IllegalStateException e) {
            return ResponseEntity.status(409).body(java.util.Map.of("error", e.getMessage()));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.unprocessableEntity().body(java.util.Map.of("error", e.getMessage()));
        } catch (RuntimeException e) {
            return ResponseEntity.notFound().build();
        }
    }

    @PatchMapping("/{id}/transitions/completed-to-active")
    public ResponseEntity<?> transitionCompletedToActive(@PathVariable Long id) {
        try {
            service.transitionCompletedToActive(id);
            return ResponseEntity.ok().build();
        } catch (IllegalStateException e) {
            return ResponseEntity.status(409).body(java.util.Map.of("error", e.getMessage()));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.unprocessableEntity().body(java.util.Map.of("error", e.getMessage()));
        } catch (RuntimeException e) {
            return ResponseEntity.notFound().build();
        }
    }

    @PatchMapping("/{id}/transitions/draw-to-active")
    public ResponseEntity<?> transitionDrawToActive(@PathVariable Long id) {
        try {
            service.transitionDrawToActive(id);
            return ResponseEntity.ok().build();
        } catch (IllegalStateException e) {
            return ResponseEntity.status(409).body(java.util.Map.of("error", e.getMessage()));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.unprocessableEntity().body(java.util.Map.of("error", e.getMessage()));
        } catch (RuntimeException e) {
            return ResponseEntity.notFound().build();
        }
    }

    @PatchMapping("/{id}/transitions/bye-to-active")
    public ResponseEntity<?> transitionBYEToActive(@PathVariable Long id) {
        try {
            service.transitionBYEToActive(id);
            return ResponseEntity.ok().build();
        } catch (IllegalStateException e) {
            return ResponseEntity.status(409).body(java.util.Map.of("error", e.getMessage()));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.unprocessableEntity().body(java.util.Map.of("error", e.getMessage()));
        } catch (RuntimeException e) {
            return ResponseEntity.notFound().build();
        }
    }
}
